src/server/accept.rs

```diff
@@ -73,47 +83,54 @@ where
             Err(_) => break ExitCode::FAILURE,
         };
 
+        let should_try_join = try_join.get();
+
         tokio::select! {
-            // Wait for the next stream from the tcp listener.
-            accepted = listener.accept() => {
-                match accepted {
-                    Err(error) => log!("error(accept): {}", error),
-                    Ok((stream, _)) => {
-                        let future = handle_connection(
+            // Check if there is a connection ready to be accepted from the
+            // listener first. This way we don't burn a permit just to join
+            // a connection.
+            biased;
+
+            // Wait for a connection from the listener
+            accepted = listener.accept() => match accepted {
+                // Spawn a task to serve the connection.
+                Ok((stream, _)) => {
+                    connections.spawn(
+                        handle_connection(
+                            permit,
```

Comment:
Not having any closures or async blocks in the accept loop is great. It also avoids an allocation on the main thread in the accept loop by keeping the task size within the `BOX_FUTURE_THRESHOLD = 2048`.